snuba/migrations/snuba_migrations/sessions/0002_sessions_aggregates.py

```diff
@@ -0,0 +1,279 @@
+    allows to query arbitrary quantiles independent of the definition.
+
+    Also, it would be nice to rename all the `countIf` `sessions_X` columns to
+    `sessions_X_count`, but that would require clickhouse 20.4.
+    """
+
+    blocking = False
+
+    def forwards_local(self) -> Sequence[operations.Operation]:
+        return [
+            # add new columns
+            operations.AddColumn(
+                storage_set=StorageSetKey.SESSIONS,
+                table_name="sessions_raw_local",
+                column=Column("quantity", WithDefault(UInt(32), str(DEFAULT_QUANTITY))),
+                after="distinct_id",
+            ),
+            operations.AddColumn(
+                storage_set=StorageSetKey.SESSIONS,
+                table_name="sessions_hourly_local",
+                column=Column(
+                    "duration_avg", AggregateFunction("avgIf", UInt(32), UInt(8))
+                ),
+                after="duration_quantiles",
+            ),
+            operations.AddColumn(
+                storage_set=StorageSetKey.SESSIONS,
+                table_name="sessions_hourly_local",
+                column=Column(
+                    "sessions_sum", AggregateFunction("sumIf", UInt(32), UInt(8))
+                ),
+                after="sessions",
+            ),
+            operations.AddColumn(
+                storage_set=StorageSetKey.SESSIONS,
+                table_name="sessions_hourly_local",
+                column=Column(
+                    "sessions_crashed_sum",
+                    AggregateFunction("sumIf", UInt(32), UInt(8)),
+                ),
+                after="sessions_crashed",
+            ),
+            operations.AddColumn(
+                storage_set=StorageSetKey.SESSIONS,
+                table_name="sessions_hourly_local",
+                column=Column(
+                    "sessions_abnormal_sum",
+                    AggregateFunction("sumIf", UInt(32), UInt(8)),
+                ),
+                after="sessions_abnormal",
+            ),
+            operations.AddColumn(
+                storage_set=StorageSetKey.SESSIONS,
+                table_name="sessions_hourly_local",
+                column=Column(
+                    "sessions_errored_sum",
+                    AggregateFunction("sumIf", UInt(32), UInt(8)),
+                ),
+                after="sessions_errored",
+            ),
+            # re-create the materialized view
+            operations.DropTable(
+                storage_set=StorageSetKey.SESSIONS,
+                table_name="sessions_hourly_mv_local",
+            ),
+            operations.CreateMaterializedView(
+                storage_set=StorageSetKey.SESSIONS,
+                view_name="sessions_hourly_mv_local",
+                destination_table_name="sessions_hourly_local",
+                columns=aggregate_columns,
+                query=query,
+            ),
+        ]
```

Comment:
I think this approach may work for self hosted environments where we can stop everything, then apply all migrations then restart the consumers, but it won't work for our production environment. In that case we cannot synchronize the migration with a deployment of a snuba PR.

So we will need to split this work into multiple PRs
1) create a backward compatible migration. After this migration we will have the new columns being filled, side by side with the old ones. The olld ones need to still be filled because the query path only knows about them. We merge that and then we can run the migration in prod.
2) run the prod migration.
3) create a second PR that changes the query path so sum the old and the new columns, at this point since the migration in prod alrleady happened, the queries will no tfail
4) now you change the processor and start filling in the new columns and stop filling in the old ones, so we are sure the query path is already in place and working in production before we start fillling in the new columns.

Reply:
That makes total sense. So:

1. add new columns
2. update the queries to consume the (empty) new columns
3. change the materialization statement to write to the new columns
4. update the processor to be able to consume the new `state` and `quantity`

Is there a way that I can atomically change the materialization query, so that I don’t under/over count?

Reply:
> Is there a way that I can atomically change the materialization query, so that I don’t under/over count?

In clickhouse you can never run multiple statements atomically, but to be fair I am not sure about what you are asking for. What are the operations you would like to execute atomically with the change to the materialization query ?